src/frontend/contexts/LocalDiscoveryContext.tsx

```diff
@@ -259,6 +278,61 @@ export function createLocalDiscoveryController(opts: {
 
 function noop() {}
 
+function startZeroconf(zeroconf: Zeroconf): Promise<void> {
+  return new Promise((resolve, reject) => {
+    const cleanup = () => {
+      zeroconf.off('start', onStart);
+      zeroconf.off('error', onError);
+    };
+    const onStart = () => {
+      cleanup();
+      resolve();
+    };
+    const onError = (err: Error) => {
+      cleanup();
+      reject(err);
+    };
+    zeroconf.on('start', onStart);
+    zeroconf.on('error', onError);
+
+    zeroconf.scan(ZEROCONF_SERVICE_TYPE, ZEROCONF_PROTOCOL, ZEROCONF_DOMAIN);
+  });
+}
+
+function stopZeroconf(zeroconf: Zeroconf): Promise<void> {
+  return new Promise((resolve, reject) => {
+    const cleanup = () => {
+      zeroconf.off('stop', onStop);
+      zeroconf.off('error', onError);
+    };
+    const onStop = () => {
+      cleanup();
+      resolve();
+    };
+    const onError = (err: Error) => {
+      cleanup();
+      reject(err);
+    };
+    zeroconf.on('stop', onStop);
+    zeroconf.on('error', onError);
+
+    zeroconf.stop();
+  });
+}
+
+function zeroconfServiceToMapeoPeer({
+  addresses,
+  port,
+  name,
+}: Readonly<ZeroconfService>): null | {
+  address: string;
+  port: number;
+  name: string;
+} {
+  const address = addresses[0];
```

Comment:
Should we do anything to handle loopback addresses? I _think_ not (that de-duplication should be handled by the `name`).